distributed/comm/mpi.py

```diff
@@ -0,0 +1,781 @@
+
+                await self.mpi_send(packed_msg, s_size, tag)
+                tag = tag + 1
+
+                # Send which frames are CUDA (bool) and
+                # how large each frame is (uint64)
+                #await self.ep.send(
+                #    struct.pack(nframes * "?" + nframes * "Q", *cuda_frames, *sizes)
+                #)
+                packed_msg = struct.pack(nframes * "?" + nframes * "Q", \
+                                                         *cuda_frames, *sizes)
+                s_size = struct.calcsize(nframes * "?" + nframes * "Q")
+
+                await self.mpi_send(packed_msg, s_size, tag)
+                tag = tag + 1
+                logger.debug("write_2: me=%s, you=%s, tag=%s", MPI.COMM_WORLD.Get_rank(), self.peer_rank, tag)
+
+                # Send frames
+
+                # It is necessary to first synchronize the default stream 
+                # before start sending. 
+
+                # non-blocking CUDA streams.
+                if any(cuda_send_frames):
+                    synchronize_stream(0)
+
+                sizes_new = tuple(i for i in sizes if i != 0) # remove all 0's from this list
+                read_counter = 0
+
+                for each_frame in send_frames:
+                    #await self.ep.send(each_frame)
+                    s_size = sizes_new[read_counter]
+                    await self.mpi_send(each_frame, s_size, tag)
+                    tag = tag + 1
+                    read_counter = read_counter + 1
+                logger.debug("write_3: me=%s, you=%s, tag=%s, sizes=%s, sizes_new=%s", MPI.COMM_WORLD.Get_rank(), self.peer_rank, tag, sizes, sizes_new)
+                return sum(sizes_new)
+            except (Exception):
+                self.abort()
+                raise CommClosedError("While writing, the connection was closed")
+
+    # a utility function for sending messages larger than CHUNK_SIZE data
+    async def mpi_send_large(self, buf, size, _tag):
+
+        me = MPI.COMM_WORLD.Get_rank()
+        you = self.peer_rank
+         
+        logger.debug("mpi_send_large: host=%s, me=%s, you=%s, tag=%s, size=%s, type(buf)=%s", \
+                                               socket.gethostname(), me, you, _tag, size, type(buf))
+
+        blk_size = CHUNK_SIZE
+        num_of_blks   = int(size / blk_size)
+        last_blk_size =     size % blk_size
+
+        logger.debug("mpi_send_large: blk_size=%s, num_of_blks=%s, last_blk_size=%s", \
+                                                          blk_size, num_of_blks, last_blk_size)
+
+        num_of_reqs = num_of_blks
+
+        if last_blk_size is not 0:
+            num_of_reqs = num_of_reqs + 1
+
+        reqs = []        
+
+        for i in range(num_of_blks):
+            s_idx = (i)   * blk_size
+            e_idx = (i+1) * blk_size
+
+            if type(buf) == rmm._lib.device_buffer.DeviceBuffer:
```

Comment:
Why do we need to check the type? This also seems to be reaching into a private module

Reply:
MPI cannot send messages larger than `2GB - 1` because it uses a signed integer to specify the number of bytes sent/received. For this reason, the MPI device needs to chunk data into blocks and communicate. 

This chunking works as long as the datatype communicated is sub-scriptable (like cuPy for example). Sub-scriptable here means using the `:` operator to index into specific regions of the buffer. However if the data being communicated is `rmm._lib.device_buffer.DeviceBuffer`, then its not sub-scriptable and hence a special case in the code. In this case, the device directly initializes a shadow buffer (of type `rmm.DeviceBuffer`) and communicate the next chunk.  

Is there a better way to do this?

Reply:
A brief search didn't find any other ways around the special case for RMM buffers.  One thought would be to convert RMM's buffer to some other type of buffer (e.g. PyBuffer, which is sub-scriptable/sliceable), but as far as I can tell, the other buffer types that I considered require extra communication and/or fully transferring the data back to CPU memory.  That would have negative impacts on performance, so I'd advice just leaving it like this.